### DISCUSSION TITLE
Documentation of access throttling feature is misleading / wrong about actual behaviour

### DISCUSSION THREAD

**Discussion by PaulBaileyAtChan** [ExpressionEngine/ExpressionEngine#4373] [Feature Requests]
I've been puzzled by the behaviour of access throttling in EE for a while. Basically, I not-infrequently trip it myself as I'm working on a site. But, based on how I understood it to work, I assumed it was pretty much impossible for anything other than a bot/spider to be throttled using the settings I have.

So I spent some time digging around in the throttling code, and it turns out I've been misunderstanding all along how it works. And, honestly, I think that's because the documentation is at best ambiguous, and actually pretty misleading / wrong. It's possible that the documentation is itself based on a misunderstanding of the code.

Here's a section of the documentation:

![Screenshot 2024-06-27 at 1 33 06 AM](https://github.com/ExpressionEngine/ExpressionEngine/assets/46760375/1dada7ec-3464-4d46-83e8-43f24600e576)

My reading of this is that there's a moving window of X seconds. If any single IP address loads too many pages within that moving window, it's blocked for a while. But that isn't at all how the throttling is implemented. And it *can't* be. To have a moving window, the time of each page-load would have to be stored; but EE only stores the time of the *last* page-load. Given how it *does* work, that's enough. Here's the relevant code (lines 105–111 of Throttling.php):

![Screenshot 2024-06-27 at 1 37 42 AM](https://github.com/ExpressionEngine/ExpressionEngine/assets/46760375/85d62dbf-4616-416c-9734-9ab353e54597)

What this does is check, for each page-load, if the gap from the previous page-load is smaller than the window of X seconds. If it is, the hit count is incremented; otherwise it's reset. Elsewhere, the code then compares the hit count against the limit, and blocks the user if the limit is exceeded.

The effect of this is that EE is checking whether a sequence of consecutive page-loads from the same IP has a gap smaller than X seconds *between each of them*. This is a very different test than whether the page-loads occur during a single window of that size, and *much* easier to trip.

How the code works seems fine enough, but using it effectively probably depends on understanding how it works, and I'm really not sure that the documentation as it stands is up to the job.

Especially since the actual access throttling is less stringent than the documentation describes, which makes it much easier to trip inadvertently (especially by developers, who might need to load multiple pages quickly!), a couple of suggested additions are:

- Allow page-loads made by logged-in EE members to be exempt from access throttling.
- Allow specific, named IP addresses to be exempt from access throttling.

**Comment by PaulBaileyAtChan** [comment]
Another issue: It looks like page-loads when **previewing** an entry from within the Entry Manager are also counted against access throttling hits. This clearly doesn't make any sense.

**Comment by intoeetive** [comment]
@robinsowell in case you've missed these notes
